pilot/pkg/security/authz/builder/builder.go

```diff
@@ -264,42 +299,62 @@ func (b Builder) buildHTTP(rule *builtRule, logger *AuthzLogger) []*hcm.HttpFilt
 		}
 	}
 
-	extauthz, err := getExtAuthz(b.extensions, providers)
-	if err != nil {
-		logger.AppendError(multierror.Prefix(err, "failed to process CUSTOM action, will generate deny configs for the specified rules:"))
-		rbac := &rbachttp.RBAC{Rules: getBadCustomDenyRules(rules)}
-		return []*hcm.HttpFilter{
-			{
+	// For CUSTOM policies, generate filter pairs for each unique provider
+	var filters []*hcm.HttpFilter
+
+	// Get unique providers and sort for deterministic output
+	uniqueProviders := maps.Keys(rule.providerRules)
+	sort.Strings(uniqueProviders)
+
+	for _, provider := range uniqueProviders {
+		providerRules := rule.providerRules[provider]
+		providerShadowRules := rule.providerShadowRules[provider]
+
+		extauthz, err := getExtAuthz(b.extensions, []string{provider})
+		if err != nil {
+			logger.AppendError(multierror.Prefix(err, fmt.Sprintf("failed to process CUSTOM action for provider %s, will generate deny configs:", provider)))
+			rbac := &rbachttp.RBAC{Rules: getBadCustomDenyRules(providerRules)}
+			filters = append(filters, &hcm.HttpFilter{
 				Name:       wellknown.HTTPRoleBasedAccessControl,
 				ConfigType: &hcm.HttpFilter_TypedConfig{TypedConfig: protoconv.MessageToAny(rbac)},
-			},
+			})
+			continue
 		}
+
+		// Add the RBAC filter in shadow mode so that it only evaluates the matching rules for CUSTOM action but not enforce it.
+		// The evaluation result is stored in the dynamic metadata keyed by the policy name. And then the ext_authz filter
+		// can utilize these metadata to trigger the enforcement conditionally.
+		// See https://docs.google.com/document/d/1V4mCQCw7mlGp0zSQQXYoBdbKMDnkPOjeyUb85U07iSI/edit#bookmark=kix.jdq8u0an2r6s
+		// for more details.
+		rbac := &rbachttp.RBAC{
+			ShadowRules:           providerRules,
+			ShadowRulesStatPrefix: authzmodel.RBACExtAuthzShadowRulesStatPrefix,
+		}
+		if providerShadowRules != nil && len(providerShadowRules.Policies) > 0 {
+			rbac.Rules = providerShadowRules
```

Comment:
doesn't this actually enable enforcement instead of shadowing for `providerShadowRules`?